unittested.d

```diff
@@ -0,0 +1,154 @@
+// D unittest runner for Codewars.
+module unittested;
+
+version(unittest):
+
+static if (!__traits(compiles, () { static import dub_test_root; })) {
+    static assert(false, "Couldn't find 'dub_test_root'. Make sure you are running tests with `dub test`");
+} else {
+    static import dub_test_root;
+}
+
+import std.stdio : stdout;
+
+shared static this() {
+    import core.runtime : Runtime, UnitTestResult;
+
+    Runtime.extendedModuleUnitTester = function () {
+        ulong passed, failed;
+
+        Test[] tests;
+        // Test discovery
+        foreach (m; dub_test_root.allModules) {
+            import std.meta : Alias;
+            import std.traits : fullyQualifiedName;
+            static if (__traits(isModule, m)) {
+                alias module_ = m;
+            } else {
+                // For cases when module contains member of the same name
+                alias module_ = Alias!(__traits(parent, m));
+            }
+
+            // Unittests in the module
+            foreach (test; __traits(getUnitTests, module_)) {
+                tests ~= Test(fullyQualifiedName!test, getTestName!test, &test);
+            }
+        }
+
+        foreach (test; tests) {
+            TestResult result;
+            scope(exit) {
+                if (result.passed) {
+                    test.onPassed(result);
+                    passed += 1;
+                } else {
+                    test.onFailed(result);
+                    failed += 1;
+                }
+                test.onEnd(result);
+            }
+            test.onStart();
+            test.run(result);
+        }
+
+        return UnitTestResult(passed + failed, passed, false, false);
+    };
+}
+
+void onStart(Test test) {
+    stdout.writefln("\n<IT::>%s", escapeLF(test.testName));
+}
+
+void onEnd(Test test, TestResult result) {
+    stdout.writefln("\n<COMPLETEDIN::>%d", result.duration);
+}
+
+void onPassed(Test test, TestResult result) {
+    stdout.writefln("\n<PASSED::>Test Passed");
+}
+
+void onFailed(Test test, TestResult result) {
+    import std.algorithm : canFind;
+
+    foreach (th; result.thrown) {
+        if (th.type == "core.exception.AssertError" || th.type == "fluentasserts.core.base.TestException") {
```

Comment:
Is there a better way to detect assertion failures?

Reply:
Why are you saving the type and not the actual exception object?

Reply:
This was derived from `silly`, and that's what they had (https://gitlab.com/AntonMeep/silly/-/blob/master/silly.d#L265).

If I keep the actual exception, is the syntax to check the type `if (cast(AssertError) exception)`?
For `fluentasserts.core.base.TestException`, don't I need that type resolvable? I want to keep that optional.